### ISSUE TITLE
High CPU load if "You are up to date" is showing (Linux)

### ISSUE THREAD

**Issue by TrueBrain** [gitbutlerapp/gitbutler#4038]
Hi there!

I noticed that GitButler was consuming a lot of CPU when it was just idling running on my Linux machine. So I tried to look into it.

The symptoms:
- When I open up a git repository that does not have any virtual branches, the CPU spikes. As soon as I create a virtual branch, the CPU drops.
- This happens for any repository I try, in any other combination. As long as there isn't a virtual branch and I have the Workspace open, it is burning CPU cycles.
- When I click on an unapplied branch, the problem also goes away. Or when I click on anything else, for that matter.
- In short, when ever I see the "You are up to date" banner, my CPU is getting hot hot hot.

I tried to use the DevTools to figure out what is happening exactly, but got a bit lost. It does correctly identify the CPU is at 100%, but it also tells me no Javascript is running, and that painting only took 7ms. The only weird thing I can see, is that `Layout & Rendering` is doing a `Composite` every 100ms. This seems excessive. But what is causing that, it isn't tell me.

Anyway, hope you can reproduce and solve this issue. Tnx!

**Comment by Byron** [comment]
Thanks for reporting! A quick test with v0.12.2 didn't show such symptoms on MacOS, but that was expected as well since that's the most well tested platform right now.

I assume you are using the same version, but would be interested to learn about the installation method. Flatpack usually means trouble still.

Right now it *seems like* it's a UI-engine issue, and to verify that you could [tail the log](https://docs.gitbutler.com/development/debugging#logs) just to be sure it's not spamming activity despite being idle. Could you also share which repo was loaded (if open-source) or roughly if it's a large repo if it's private?

Maybe there is also a frontend profile file you could share. Maybe @ndom91 also has an idea what it could be (or why that doesn't seem to be a problem on MacOS).

Thanks a lot for your help with this.

**Comment by TrueBrain** [comment]
> I assume you are using the same version, but would be interested to learn about the installation method. Flatpack usually means trouble still.

I should have mentioned that :) v0.12.2, installed via Debian package.

> could [tail the log](https://docs.gitbutler.com/development/debugging#logs) just to be sure it's not spamming activity despite being idle

Nothing is being written in the logs while idle (and with high CPU).

> Could you also share which repo was loaded (if open-source) or roughly if it's a large repo if it's private?

```
mkdir test
cd test
git init
git commit -m "chore: initial empty commit" --allow-empty
git remote add origin file:///<something>
```

That already does the trick :)

I hope one of you can reproduce it; or it might just be my setup, ofc. If you need any more info, let me know!

**Comment by Byron** [comment]
Thanks a lot!

Then it looks like a frontend issue, and I assume it's the renderer. Since dev-mode works I wonder if you have access to profiling tools that can output profile-files to share with us? Alternatively, a screen-recording showing the UI could also help to spot things.

My hope would be that indeed there is an application-controlled action happening every 100ms which could be adjusted to fix the issue, after-all it seems fine when *something* is rendered.

**Comment by TrueBrain** [comment]
A screen-recording is identical to a still image, so I can do that :)

![image](https://github.com/gitbutlerapp/gitbutler/assets/1663690/1bf70cbd-70fd-480d-b68a-26d329f4fc7d)

I can also make a profile via the dev-tools, which looks something like this:

![image](https://github.com/gitbutlerapp/gitbutler/assets/1663690/60d50336-2e1d-4203-8db6-878e89ca8334)

I also exported the profile, in case that is useful (but it is really boring, as it kinda only shows that it is redrawing things every 100ms):

[localhost-recording.json.zip](https://github.com/user-attachments/files/15749622/localhost-recording.json.zip)

As for an actual profile, as I don't have the debug symbols, it is not all that useful. All you see is random pointers :D The only thing that jumps out, that it is `libpixman` doing the actual work. But as to how it got there ... Screenshot to show my point:

![image](https://github.com/gitbutlerapp/gitbutler/assets/1663690/ef6aa94d-ea19-4bf3-9c1d-0b7f7b44a709)

If you have suggestions how to make a better profile, do let me know!

**Comment by Byron** [comment]
Thanks so much! I found the screenshot of the callstack particularly interesting. I think it shows the app after startup, and it doesn't look like the high-CPU loop is really shown there. Maybe it's in another process. But I could get it all wrong as well as the actual time isn't shown, just percentages.

> I also exported the profile, in case that is useful (but it is really boring, as it kinda only shows that it is redrawing things every 100ms):

Thanks a lot for that!

What I did find interesting is that the HighCPU usage 'only' happens for the first  2 seconds, which certainly is impressive on a mostly empty repository. It's better than 'all the time' though which the subject of the issue currently suggests.

With that said, I still think it's worth finding out what happens in the first two seconds in terms of drawing. In my own profiled testing, I noticed quite some activity when changing to a new project, but it never was high CPU usage for a long time, but fizzled out over more or less 2 seconds. 

Maybe that time is related to initial checks… but those run mostly in the backend while the frontend should have nothing to do if there is nothing to display.

Still puzzled….
<img width="539" alt="Screenshot 2024-06-09 at 06 50 17" src="https://github.com/gitbutlerapp/gitbutler/assets/63622/ee503ffa-da17-46e2-a897-37a0a1d051cf">

**Comment by TrueBrain** [comment]
The profile I gave is only if two seconds. The high CPU never stops for me, and just keeps on going. Even after 5 minutes it was still this high.

So sadly, nothing to do with initialization :(

**Comment by Byron** [comment]
I see :/, thanks for the clarification.

**Comment by TrueBrain** [comment]
As I could reproduce this problem easily with a debug build, it also didn't take long to find the culprit: the animation of the zen computer, hovering slowly up&down. Looking back, that of course makes total sense. It is nearly invisible that the image moves up and down constantly, but this updates every 100ms, causing the crazy CPU load. So I was wrong in an earlier statement, that a screenshot would be the same as a movie. There are a few pixels constantly being changed :)

To be very exact, the issue is this line:

https://github.com/gitbutlerapp/gitbutler/blob/ea31b78f6a6950381e8e07f8c95d7ac375201d05/app/src/lib/components/Board.svelte#L294

Disabling that, instantly fixes the CPU load.

I am not sure what a good resolution would be. I can fully understand that such an animation is fun and adds a bit extra to the Zen. On the other hand, it really is not worth burning a CPU core for constant redrawing :) But I better leave the solution part to you guys; you most likely know better anyway :)

**Comment by Byron** [comment]
> I am not sure what a good resolution would be. I can fully understand that such an animation is fun and adds a bit extra to the Zen.

😂You made my day, thank you :D.

<img width="1422" alt="Screenshot 2024-06-11 at 08 06 46" src="https://github.com/gitbutlerapp/gitbutler/assets/63622/55b8e7d9-e4c5-4b59-b896-9f45d558006b">

It's great you found the culprit, thank for your help with this! There must be some optimization missing as on MacOS, it's not showing up on the CPU meter. Maybe it's all GPU there and it doesn't care.

Nonetheless, it's a performance issue and I'd hope the computer can just be as zen as the user when this picture shows up :). I think @PavelLaptev might know what to do here. Maybe a setting that reduces animation for platforms that have trouble with it?

PS: I never noticed the movement until I looked closely. Definitely a nice touch though :).

**Comment by PavelLaptev** [comment]
@Byron @TrueBrain thanks for reporting this.
Let me also check it on my end as well.

**Comment by ndom91** [comment]
Yeah interesting, that's only animating the `transform` property which should be fairly performant, and you're already using the `translateZ(0)` "hack" :thinking: 

```css
.empty-board__image {
	position: absolute;
	top: 50%;
	left: 50%;
	transform: translate(-50%, -70%) translateZ(0);
	width: 13.3rem;
	animation: hovering 5.5s infinite ease-in-out;
	animation-delay: 3s;
}

@keyframes hovering {
	0% {
		transform: translate(-50%, -70%) translateZ(0);
	}
	50% {
		transform: translate(-50%, -65%) translateZ(0);
	}
	100% {
		transform: translate(-50%, -70%) translateZ(0);
	}
}
```

@PavelLaptev I can take a look on linux tomorrow as well if you can't reproduce it on OS X.

**Comment by PavelLaptev** [comment]
@ndom91 sounds good, thank you
